File: src/python/openloop.py
"""OpenLoop 运行时模块

在 Pyodide 中注册为 'openloop'，提供：
  - parameter(name, default, *, min, max, step) → 注册并返回用户可调参数
  - status(name, value) → 将变量发送到检查器/图表

用户通过 import openloop 或 import openloop as ol 使用。
"""

_params = {}   # {name: value} — 运行时值，TS 侧通过 _set_params 注入
_status = {}   # {name: value} — 运行时状态值


def parameter(name, default, *, min=None, max=None, step=None):
    """注册并返回一个用户可调参数。

    Args:
        name: 参数名（字符串）
        default: 默认值（数值）
        min: 最小值（可选，给 PanelParams 用）
        max: 最大值（可选，给 PanelParams 用）
        step: 步长（可选，给 PanelParams 用）

    Returns:
        当前参数值（首次调用返回 default，之后返回滑块值）
    """
    if name not in _params:
        _params[name] = default
    return _params[name]


def status(name, value):
    """将一个变量发送到检查器/图表。

    Args:
        name: 变量名（字符串）
        value: 当前值
    """
    _status[name] = value


# ── 辅助函数（TS 侧调用） ──

def _get_params():
    """获取所有参数当前值。"""
    return dict(_params)


def _set_params(d):
    """批量设置参数值（TS 侧从滑块注入）。"""
    _params.update(d)


def _get_status():
    """获取所有状态值。"""
    return dict(_status)


def _clear():
    """清空参数和状态（仿真开始/停止时调用）。"""
    _params.clear()
    _status.clear()

File: src/modules/python/openloop.ts
import { usePyodide } from './pyodide'
import openloopCode from '@/python/openloop.py?raw'

const pyodide = usePyodide()
let injected = false

/**
 * 注入 openloop 模块到 Pyodide 全局命名空间。
 * 在仿真启动前调用一次。
 * 失败时 throw Error，让调用方能正确中止。
 */
export async function injectOpenLoop() {
  if (injected) return
  if (!pyodide.isReady.value) {
    await pyodide.init()
  }
  if (!pyodide.isReady.value) {
    throw new Error('Pyodide 未就绪，无法注入 openloop 模块')
  }

  // 执行 openloop.py，定义函数到 globals
  const r1 = pyodide.runPython(openloopCode)
  if (r1.error) {
    throw new Error(`openloop 模块注入失败: ${r1.error}`)
  }

  // 创建 openloop 模块对象并注册到 sys.modules
  const r2 = pyodide.runPython(`
import sys as _ol_sys, types as _ol_types
openloop = _ol_types.ModuleType('openloop')
openloop.parameter = parameter
openloop.status = status
_ol_sys.modules['openloop'] = openloop
del _ol_types
`)
  if (r2.error) {
    throw new Error(`openloop 模块注册失败: ${r2.error}`)
  }

  injected = true
}

/** 将滑块值写入 openloop 的 _params dict */
export function updateParamValues(values: Record<string, number>) {
  const py = pyodide.pyodide.value
  if (!py) return
  const setParams = py.globals.get('_set_params')
  if (!setParams) return
  const pyDict = py.toPy(values)
  setParams(pyDict)
  pyDict.destroy()
  setParams.destroy()
}

/** 读取 openloop 的 _status dict */
export function getStatusValues(): Record<string, number> {
  const py = pyodide.pyodide.value
  if (!py) return {}
  const getStatus = py.globals.get('_get_status')
  if (!getStatus) return {}
  const result = getStatus()
  const jsResult = result.toJs({ dict_converter: Object.fromEntries }) as Record<string, number>
  result.destroy()
  getStatus.destroy()
  return jsResult
}

/** 清空参数和状态 */
export function clearOpenLoop() {
  const py = pyodide.pyodide.value
  if (!py) return
  const clear = py.globals.get('_clear')
  if (!clear) return
  clear()
  clear.destroy()
}

/** 重置注入状态（测试用） */
export function resetOpenLoopModule() {
  injected = false
}

export function useOpenLoopModule() {
  return {
    injectOpenLoop,
    updateParamValues,
    getStatusValues,
    clearOpenLoop,
    resetOpenLoopModule,
  }
}
